tools/Environments/DevHome.Environments/ViewModels/PerProviderViewModel.cs

```diff
@@ -0,0 +1,139 @@
+        ComputeSystems = new ObservableCollection<ComputeSystemCardBase>(computeSystems);
+        _mainWindow = mainWindow;
+
+        _stringResource = new StringResource("DevHome.Environments.pri", "DevHome.Environments/Resources");
+        ComputeSystemAdvancedView = new AdvancedCollectionView(ComputeSystems);
+        ComputeSystemAdvancedView.SortDescriptions.Add(new SortDescription("IsCardCreating", SortDirection.Descending));
+    }
+
+    /// <summary>
+    /// Updates the view model to show only the compute systems that match the search criteria.
+    /// </summary>
+    public void SearchHandler(string query)
+    {
+        ComputeSystemAdvancedView.Filter = system =>
+        {
+            if (system is CreateComputeSystemOperationViewModel createComputeSystemOperationViewModel)
+            {
+                return createComputeSystemOperationViewModel.EnvironmentName.Contains(query, StringComparison.OrdinalIgnoreCase);
+            }
+
+            if (system is ComputeSystemViewModel computeSystemViewModel)
+            {
+                var systemName = computeSystemViewModel.ComputeSystem!.DisplayName.Value;
+                var systemAltName = computeSystemViewModel.ComputeSystem.SupplementalDisplayName.Value;
+                return systemName.Contains(query, StringComparison.OrdinalIgnoreCase) || systemAltName.Contains(query, StringComparison.OrdinalIgnoreCase);
+            }
+
+            return false;
+        };
+
+        _mainWindow.DispatcherQueue.EnqueueAsync(() =>
+        {
+            IsVisible = ComputeSystemAdvancedView.Count > 0;
+        });
+    }
+
+    /// <summary>
+    /// Updates the view model to sort the compute systems according to the sort criteria.
+    /// </summary>
+    /// <remarks>
+    /// New SortDescription property names should be added as new properties to <see cref="ComputeSystemCardBase"/>
+    /// </remarks>
+    public void SortHandler(int selectedSortIndex)
+    {
+        ComputeSystemAdvancedView.SortDescriptions.Clear();
+        switch (selectedSortIndex)
+        {
+            case 0:
+                ComputeSystemAdvancedView.SortDescriptions.Add(new SortDescription("Name", SortDirection.Ascending));
+                break;
+            case 1:
+                ComputeSystemAdvancedView.SortDescriptions.Add(new SortDescription("Name", SortDirection.Descending));
+                break;
+            case 2:
+                ComputeSystemAdvancedView.SortDescriptions.Add(new SortDescription("LastConnected", SortDirection.Ascending));
+                break;
+        }
+    }
+
+    /// <summary>
+    /// Updates the view model to filter the compute systems according to the provider.
+    /// </summary>
+    public void ProviderHandler(string currentProvider)
+    {
+        ComputeSystemAdvancedView.Filter = system =>
+        {
+            if (currentProvider.Equals(_stringResource.GetLocalized("AllProviders"), StringComparison.OrdinalIgnoreCase))

```

Comment:
Place this check above `ComputeSystemAdvnaceView...`  Check early.

Reply:
Can't. Has to be part of the lambda filter function.

Reply:
After looking at the other comments, I understand what is happening here.  This is to return true if the filter is set to "AllProviders" and the check needs to be here.

Gotcha.